pkg/db/v1beta1/postgres/postgres.go

```diff
@@ -0,0 +1,226 @@
+	dbName := env.GetEnvOrDefault(common.DefaultPostgreSQLDatabase,
+		common.DefaultPostgreSQLDatabase)
+
+	psqlInfo := fmt.Sprintf("host=%s port=%s user=%s "+
+		"password=%s dbname=%s sslmode=disable",
+		dbHost, dbPort, dbUser, dbPass, dbName)
+
+	return psqlInfo
+}
+
+func NewDBInterface() (common.KatibDBInterface, error) {
+	db, err := openSQLConn(dbDriver, getDbName(), connectInterval, connectTimeout)
+	if err != nil {
+		return nil, fmt.Errorf("DB open failed: %v", err)
+	}
+	return NewWithSQLConn(db)
+}
+
+func openSQLConn(driverName string, dataSourceName string, interval time.Duration,
+	timeout time.Duration) (*sql.DB, error) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	timeoutC := time.After(timeout)
+	for {
+		select {
+		case <-ticker.C:
+			if db, err := sql.Open(driverName, dataSourceName); err == nil {
+				// if db, err := sql.Open(driverName, dataSourceName); err == nil {
+				if err = db.Ping(); err == nil {
+					return db, nil
+				}
+				klog.Errorf("Ping to Katib db failed: %v", err)
+			} else {
+				klog.Errorf("Open sql connection failed: %v", err)
+			}
+		case <-timeoutC:
+			return nil, fmt.Errorf("Timeout waiting for DB conn successfully opened.")
+		}
+	}
+}
+
+func NewWithSQLConn(db *sql.DB) (common.KatibDBInterface, error) {
+	d := new(dbConn)
+	d.db = db
+	seed, err := crand.Int(crand.Reader, big.NewInt(1<<63-1))
+	if err != nil {
+		return nil, fmt.Errorf("RNG initialization failed: %v", err)
+	}
+	// We can do the following instead, but it creates a locking issue
+	//d.rng = rand.New(rand.NewSource(seed.Int64()))
+	rand.Seed(seed.Int64())
+
+	return d, nil
+}
+
+func (d *dbConn) RegisterObservationLog(trialName string, observationLog *v1beta1.ObservationLog) error {
+	statement := "INSERT INTO observation_logs (trial_name, time, metric_name, value) VALUES "
+	values := []interface{}{}
+
+	i := 1
+	for _, mlog := range observationLog.MetricLogs {
+		if mlog.TimeStamp == "" {
+			continue
+		}
+		t, err := time.Parse(time.RFC3339Nano, mlog.TimeStamp)
+		if err != nil {
+			return fmt.Errorf("Error parsing start time %s: %v", mlog.TimeStamp, err)
+		}
+		sqlTimeStr := t.UTC().Format(time.RFC3339Nano)
+
+		statement += fmt.Sprintf("($%d, $%d, $%d, $%d),", i, i+1, i+2, i+3)
+		values = append(values, trialName, sqlTimeStr, mlog.Metric.Name, mlog.Metric.Value)
+		i += 4
+	}
+
+	statement = statement[:len(statement)-1]
+
+	// Prepare the statement
+	stmt, err := d.db.Prepare(statement)
+	if err != nil {
+		return fmt.Errorf("Prepare SQL statement failed: %v", err)
+	}
+
+	// Defer Close the statement
+	defer stmt.Close()
+
+	// Execute INSERT
+	_, err = stmt.Exec(values...)
+	if err != nil {
+		return fmt.Errorf("Execute SQL INSERT failed: %v", err)
+	}
+
+	return nil
+}
+
+func (d *dbConn) GetObservationLog(trialName string, metricName string, startTime string, endTime string) (*v1beta1.ObservationLog, error) {
+	qfield := []interface{}{trialName}
+	qstr := ""
+
+	i := 2
```

Comment:
Since it is hard to recognize what this variable means, I'll change the entire logic to create the statement.
After reviewing the core part.

Reply:
DONE